cumulus/client/bootnodes/src/advertisement.rs

```diff
@@ -0,0 +1,592 @@
+				Protocol::Ip6(ip) => IpNetwork::from(ip).is_global(),
+				_ => true,
+			})
+		};
+
+		// Check if the address is a loopback address.
+		let is_loopback = |address: &Multiaddr| {
+			address.iter().any(|protocol| match protocol {
+				Protocol::Ip4(ip) => IpNetwork::from(ip).is_loopback(),
+				Protocol::Ip6(ip) => IpNetwork::from(ip).is_loopback(),
+				_ => false,
+			})
+		};
+
+		// 1) public addresses provided by the operator
+		let public_addresses = self.public_addresses.clone().into_iter();
+
+		// 2) global listen addresses
+		let global_listen_addresses =
+			self.parachain_network.listen_addresses().into_iter().filter(is_global);
+
+		// 3a) discovered external addresses (global)
+		let global_external_addresses =
+			self.parachain_network.external_addresses().into_iter().filter(is_global);
+
+		// 3b) discovered external addresses (non-global)
+		let non_global_external_addresses = self
+			.parachain_network
+			.external_addresses()
+			.into_iter()
+			.filter(|addr| !is_global(addr));
+
+		// 4) non-global listen addresses
+		let non_global_listen_addresses = self
+			.parachain_network
+			.listen_addresses()
+			.into_iter()
+			.filter(|addr| !is_global(addr) && !is_loopback(addr));
+
+		// 5) loopback listen addresses
+		let loopback_listen_addresses =
+			self.parachain_network.listen_addresses().into_iter().filter(is_loopback);
+
+		let mut seen = HashSet::new();
+
+		public_addresses
+			.chain(global_listen_addresses)
+			.chain(global_external_addresses)
+			.chain(
+				self.advertise_non_global_ips
+					.then_some(
+						non_global_external_addresses
+							.chain(non_global_listen_addresses)
+							.chain(loopback_listen_addresses),
+					)
+					.into_iter()
+					.flatten(),
+			)
+			.filter_map(without_p2p)
+			// Deduplicate addresses.
+			.filter(|addr| seen.insert(addr.clone()))
+			.take(MAX_ADDRESSES)
+			.collect()
+	}
+
+	fn handle_request(&mut self, req: IncomingRequest) {
+		if req.payload == self.para_id_scale_compact {
+			trace!(
+				target: LOG_TARGET,
+				"Serving paranode addresses request from {:?} for parachain ID {}",
+				req.peer,
+				self.para_id,
+			);
+
+			let response = crate::schema::Response {
+				peer_id: self.parachain_network.local_peer_id().to_bytes(),
+				addrs: self.paranode_addresses().iter().map(|a| a.to_vec()).collect(),
+				genesis_hash: self.parachain_genesis_hash.clone(),
+				fork_id: self.parachain_fork_id.clone(),
+			};
+
+			let _ = req.pending_response.send(OutgoingResponse {
+				result: Ok(response.encode_to_vec()),
+				reputation_changes: Vec::new(),
+				sent_feedback: None,
+			});
+		} else {
+			let payload = req.payload;
+			match Compact::<ParaId>::decode(&mut &payload[..]) {
+				Ok(para_id) => {
+					trace!(
+						target: LOG_TARGET,
+						"Ignoring request for parachain ID {} != self parachain ID {} from {:?}",
+						para_id.0,
+						self.para_id,
+						req.peer,
+					);
+				},
+				Err(e) => {
+					trace!(
```

Comment:
nit: Similar to discovery maybe we should decrease the peer reputation

Reply:
Yeah, the same motivation here: `/paranode` request-response protocol is not essential to the relay chain node and shouldn affect it operation.